builtin/logical/pki/cert_util.go

```diff
@@ -162,6 +241,291 @@ func validateCommonNames(req *logical.Request, commonNames []string, role *roleE
 	return "", nil
 }
 
+func generateCert(b *backend,
+	role *roleEntry,
+	signingBundle *caInfoBundle,
+	isCA bool,
+	req *logical.Request,
+	data *framework.FieldData) (*certutil.ParsedCertBundle, error) {
+
+	creationBundle, err := generateCreationBundle(b, role, signingBundle, nil, req, data)
+	if err != nil {
+		return nil, err
+	}
+
+	if isCA {
+		creationBundle.IsCA = isCA
+	}
+
+	parsedBundle, err := createCertificate(creationBundle)
+	if err != nil {
+		return nil, err
+	}
+
+	return parsedBundle, nil
+}
+
+// N.B.: This is only meant to be used for generating intermediate CAs.
+// It skips some sanity checks.
+func generateCSR(b *backend,
+	role *roleEntry,
+	signingBundle *caInfoBundle,
+	req *logical.Request,
+	data *framework.FieldData) (*certutil.ParsedCSRBundle, error) {
+
+	creationBundle, err := generateCreationBundle(b, role, signingBundle, nil, req, data)
+	if err != nil {
+		return nil, err
+	}
+
+	parsedBundle, err := createCSR(creationBundle)
+	if err != nil {
+		return nil, err
+	}
+
+	return parsedBundle, nil
+}
+
+func signCert(b *backend,
+	role *roleEntry,
+	signingBundle *caInfoBundle,
+	isCA bool,
+	useCSRValues bool,
+	req *logical.Request,
+	data *framework.FieldData) (*certutil.ParsedCertBundle, error) {
+
+	csrString := data.Get("csr").(string)
+	if csrString == "" {
+		return nil, certutil.UserError{Err: fmt.Sprintf("\"csr\" is empty")}
+	}
+
+	pemBytes := []byte(csrString)
+	pemBlock, pemBytes := pem.Decode(pemBytes)
+	if pemBlock == nil {
+		return nil, certutil.UserError{Err: "csr contains no data"}
+	}
+	csr, err := x509.ParseCertificateRequest(pemBlock.Bytes)
+	if err != nil {
+		return nil, certutil.UserError{Err: "certificate request could not be parsed"}
+	}
+
+	creationBundle, err := generateCreationBundle(b, role, signingBundle, csr, req, data)
+	if err != nil {
+		return nil, err
+	}
+
+	if isCA {
+		creationBundle.IsCA = isCA
+		creationBundle.UseCSRValues = useCSRValues
+	}
+
+	parsedBundle, err := signCertificate(creationBundle, csr)
+	if err != nil {
+		return nil, err
+	}
+
+	return parsedBundle, nil
+}
+
+func generateCreationBundle(b *backend,
+	role *roleEntry,
+	signingBundle *caInfoBundle,
+	csr *x509.CertificateRequest,
+	req *logical.Request,
+	data *framework.FieldData) (*certCreationBundle, error) {
+	var err error
+
+	// Get the common name(s)
+	cn := data.Get("common_name").(string)
+	if len(cn) == 0 {
```

Comment:
We should just fix this at a higher level instead of introducing this type of logic